controllers/scopeinstance_controller.go

```diff
@@ -58,47 +67,166 @@ func (r *ScopeInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+						// 	Name:       in.GetObjectMeta().GetName(),
+						// 	UID:        "d9607e19-f88f-11e6-a518-42010a800195",
+						// }},
 					},
 					Subjects: cr.Subjects,
 					RoleRef: rbacv1.RoleRef{
 						Kind: "ClusterRole",
 						Name: cr.GenerateName,
 					},
 				})
+
+			} else {
+				// delete from Map(key, value) if already created Cluster Role Bindings
+				delete(clusterRoleBindingMap, cr.GenerateName)
+				fmt.Println("Key found value is: ", val)
+			}
+		}
+	} else {
+
+		// it will iterate over the namespace and createrole bindings for each cluster roles
+		for _, namespace := range in.Spec.Namespaces {
+			for _, cr := range st.Spec.ClusterRoles {
+				variable := namespace + "-" + cr.GenerateName
+				fmt.Println("variable found value is here: ", variable)
+				if val, ok := roleBindingMap[variable]; !ok {
+					objects = append(objects, &rbacv1.RoleBinding{
+						ObjectMeta: metav1.ObjectMeta{
+							GenerateName: cr.GenerateName + "-",
+							Namespace:    namespace,
+							Labels: map[string]string{
+								"labels": "scopeInstanceController",
+							},
+							// OwnerReferences: []metav1.OwnerReference{{
+							// 	APIVersion: in.APIVersion,
+							// 	Kind:       in.Kind,
+							// 	Name:       in.GetObjectMeta().GetName(),
+							// 	UID:        "d9607e19-f88f-11e6-a518-42010a800195",
+							// }},
+						},
+						Subjects: cr.Subjects,
+						RoleRef: rbacv1.RoleRef{
+							Kind: "ClusterRole",
+							Name: cr.GenerateName,
+						},
+					})
+
+				} else {
+					// delete from Map(key, value) if already created Role Bindings
+					fmt.Println("Key found value is: ", val)
+					delete(roleBindingMap, variable)
+				}
 			}
 		}
+	}
+
+	// below is for cleanup, check if Map(Key, value) of Role Binding is empty or not if not then those
+	// are remaining ones which we have to delete
+	if len(roleBindingMap) != 0 {
+		for key, element := range roleBindingMap {
+			split := strings.Split(key, "-")
+
+			deleteObjects = append(objects, &rbacv1.RoleBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      element,
+					Namespace: split[0],
+					Labels: map[string]string{
+						"labels": "scopeInstanceController",
+					},
+				},
+			})
+		}
+	}
+
+	// below is for cleanup, check if Map(Key, value) of Cluster Role Binding is empty or not if not then those
+	// are remaining ones which we have to delete
+	if len(clusterRoleBindingMap) != 0 {
+		for _, element := range clusterRoleBindingMap {
+			deleteObjects = append(objects, &rbacv1.ClusterRoleBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: element,
+					Labels: map[string]string{
+						"labels": "scopeInstanceController",
+					},
+				},
+			})
+		}
+	}
 
+	// Delete Unnecessary Objects
+	for _, object := range deleteObjects {
+		log.Log.Info("Deleting object", "object", object)
+		if err := r.Client.Delete(ctx, object); err != nil {
+			log.Log.Error(err, "Error deleting object", "object", object)
+		}
 	}
 
+	// Create Objects
```

Comment:
This section is a bit messy, I think it would be clearer if we did the following for both roleBindings and clusterRoleBindings:
1. We get the list of desired (cluster)RoleBindings from the desiredResources function above.
2. We check if the list of desired (cluster)RoleBindings is of length 0. If it is, we delete the existing (cluster)RoleBindings and skip the following steps
3. We iterate over the list of desired roleBindings. If found in the map, the two objects are compared. If a diff is found, the object is updated.  If the object does not exist in the map, the object is created. The object is then removed from the map.
4. After all desired (cluster)RoleBindings have been created/updated and removed from the map, delete remaining items in the map.

Reply:
Added update logic.